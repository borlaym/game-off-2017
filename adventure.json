[
	{
	"id": "starter",
	"text": "Fury Station is a crowded cesspool today. Just like any other day.\nThe sound of distant PA announcements echo through the hall, and the chatter in dozens of languages merge into a soft, comforting murmur.\nThe line at the employment office is long enough that you start daydreaming about better times.",
	"inGameTimestamp": "2100-01-01 00:00:01",
	"options": [
		{
			"id": 1,
			"text": "Growing up on Earth was nice… at least before the radiation sickness took Mom and Dad.",
			"resultingAction": {
				"id": 1,
				"targetNode": "starter-2",
				"effects": {
					"gainTags": [{
						"target": "Self",
						"tag": "BORN ON EARTH"
					}],
					"loseTags": []
				}
			}
		},
		{
			"id": 2,
			"text": "It was fun living here on Fury Station when I was a kid. The Divine Order was strict, but at least they kept the place tidy and safe.",
			"resultingAction": {
				"id": 2,
				"targetNode": "starter-2",
				"effects": {
					"gainTags": [
						{
							"target": "Self",
							"tag": "RAISED BY DIVINE ORDER"
						}
					],
					"loseTags": []
				}
			}
		},
		{
			"id": 3,
			"text": "I’m glad I was born on a ship. Back then, we were free, the solar wind at our backs.",
			"resultingAction": {
				"id": 3,
				"targetNode": "starter-2",
				"effects": {
					"gainTags": [
						{
							"target": "Self",
							"tag": "BORN ON A SHIP"
						}
					],
					"loseTags": []
				}
			}
		}
	]},
	{
		"id": "starter-2",
		"text": "NPC: Buddy. Hey! You’re up.\nCLERK LADY: You! Yeah,[short] [guy]. Stop wasting my time and get over here.\nYou walk over to the dirty desk and sit down on the uncomfortable, creaky stool. The clerk lady has wrinkles that only decades of constant grumpiness can explain. Her raspy voice is almost unintelligible through the speakers.\nCLERK LADY: Name.",
		"inGameTimestamp": "2100-01-01 00:02:01",
		"options": [{
			"id": 1,
			"text": "<%=o.self.name%>",
			"resultingAction": {
				"id": 1,
				"targetNode": "starter-3"
			}
		}]
	},
	{
		"id": "starter-3",
		"text": "I hope I get assigned to something good.",
		"inGameTimestamp": "2100-01-01 00:02:02",
		"options": []
	}
]