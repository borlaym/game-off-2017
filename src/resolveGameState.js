// @flow

import type { Party, Character, Adventure, GameState, Tag, SaveStep, Option, ResultingAction } from './types';
import findLast from 'lodash/findLast';

const resolveGameState = ({
	adventure,
	party,
	playerId,
}: {
	adventure: Adventure,
	party: Party,
	playerId: string
}): GameState => {
	let globalTags: Array<Tag> = [];
	const resolvedCharacters: Array<Character> = JSON.parse(JSON.stringify(party.participants));
	let currentNode: Node = adventure.starter;
	const log = [];
	const player: ?Character = resolvedCharacters.find(_ => _._uid === playerId);
	if (!player) {
		throw new Error('Player not found');
	}
	Object.keys(party.save).forEach((saveKey: string) => {
		// Iterate through all save steps
		const saveStep: SaveStep = party.save[saveKey];
		const node = adventure[saveStep._nodeRef];
		const actingPlayer: ?Character =
			resolvedCharacters.find(_ => _._uid === saveStep._characterRef);
		if (!actingPlayer) {
			throw new Error('Player not found');
		}
		const allActions = node.options.reduce((acc, option) => [
			...acc,
			option.resultingAction,
		], []);
		const takenAction: ?ResultingAction = allActions.find(_ => _.id === saveStep._actionID);
		if (!takenAction) {
			return;
		}
		const takenOption: ?Option = node.options.find(_ => _.resultingAction.id === takenAction.id);
		if (!takenOption) {
			return;
		}
		// Apply effects
		if (takenAction.effects) {
			(takenAction.effects.gainTags || []).forEach((effect) => {
				if (effect.target.toLowerCase() === 'self') {
					actingPlayer.tags.push(effect.tag);
				} else {
					globalTags.push(effect.tag);
				}
			});
		}
		if (takenAction.effects) {
			(takenAction.effects.loseTags || []).forEach((effect) => {
				if (effect.target.toLowerCase() === 'self') {
					actingPlayer.tags = actingPlayer.tags.filter(_ => _ !== effect.tag);
				} else {
					globalTags = globalTags.filter(_ => _ !== effect.tag);
				}
			});
		}
		// Set next node if the action was taken by the current player
		if (actingPlayer === player) {
			currentNode = adventure[takenAction.targetNode];
			const timestamp = (new Date(currentNode.inGameTimestamp)).getTime();
			// Add message to log
			log.push({
				text: node.text,
				timestamp,
			});
			log.push({
				text: takenOption.logText || takenOption.text,
				timestamp: timestamp + 1,
			});
		}
	});
	return {
		globalTags,
		characters: resolvedCharacters,
		player,
		currentNode,
		log,
	};
};

export default resolveGameState;
