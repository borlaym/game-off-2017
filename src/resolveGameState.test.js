// @flow

import resolveGameState from './resolveGameState';
import adventure from './adventure.json';

const defaultValue = {
	adventure,
	save: {},
	participants: [{
		_uid: 'player1',
		name: 'Player One',
		tags: [],
		_partyRef: 'party1',
	}, {
		_uid: 'player2',
		name: 'Player Two',
		tags: [],
		_partyRef: 'party1',
	}],
	playerId: 'player1',
};

/**
 * Helper for quickly creating default inputs for resolveGameState
 * with provided save steps
 */
const valueWithSave = save => ({
	...defaultValue,
	save,
});

describe('resolveGameState', () => {
	it('returns the base game state', () => {
		const gameState = resolveGameState(defaultValue);
		expect(gameState).toMatchSnapshot();
	});

	it('resolves an action that gives a tag', () => {
		const gameState = resolveGameState(valueWithSave({
			save1: {
				_nodeRef: 'starter',
				_characterRef: 'player1',
				_actionID: '2',
			},
		}));
		expect(gameState.characters).toMatchSnapshot();
		expect(gameState.currentNode.id).toMatchSnapshot();
	});

	it('resolves an action that gives a global tag as well', () => {
		const gameState = resolveGameState(valueWithSave({
			save1: {
				_nodeRef: 'starter',
				_characterRef: 'player1',
				_actionID: '1',
			},
		}));
		expect(gameState.characters).toMatchSnapshot();
		expect(gameState.globalTags).toMatchSnapshot();
	});

	it('current node is calculated for the current player, not for all players', () => {
		const gameState = resolveGameState(valueWithSave({
			save1: {
				_nodeRef: 'starter',
				_characterRef: 'player2', // Note that in this case it's player 2 who takes the action
				_actionID: '1',
			},
		}));
		expect(gameState.currentNode.id).toMatchSnapshot();
	});

	it('correct log is calculated', () => {
		const gameState = resolveGameState(valueWithSave({
			save1: {
				_nodeRef: 'starter',
				_characterRef: 'player2',
				_actionID: '2',
			},
			save2: {
				_nodeRef: 'starter-2',
				_characterRef: 'player2',
				_actionID: '1',
			},
			save3: {
				_nodeRef: 'starter',
				_characterRef: 'player1',
				_actionID: '1',
			},
			save4: {
				_nodeRef: 'starter-2',
				_characterRef: 'player1',
				_actionID: '1',
			},
			save5: {
				_nodeRef: 'starter-3',
				_characterRef: 'player2',
				_actionID: '1',
			},
		}));
		expect(gameState.log).toMatchSnapshot();
	});
});
