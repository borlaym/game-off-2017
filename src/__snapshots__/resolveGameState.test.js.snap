// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveGameState correct log is calculated based on what the player can see 1`] = `
Array [
  Object {
    "text": "Fury Station is a crowded cesspool today. Just like any other day.
The sound of distant PA announcements echo through the hall, and the chatter in dozens of languages merge into a soft, comforting murmur.
The line at the employment office is long enough that you start daydreaming about better times.",
    "timestamp": 4102441321000,
  },
  Object {
    "text": "Growing up on Earth was nice… at least before the radiation sickness took Mom and Dad.",
    "timestamp": 4102441321001,
  },
  Object {
    "text": "NPC: Buddy. Hey! You’re up.
CLERK LADY: You! Yeah,[short] [guy]. Stop wasting my time and get over here.
You walk over to the dirty desk and sit down on the uncomfortable, creaky stool. The clerk lady has wrinkles that only decades of constant grumpiness can explain. Her raspy voice is almost unintelligible through the speakers.
CLERK LADY: Name.",
    "timestamp": 4102441322000,
  },
  Object {
    "text": "You say your name",
    "timestamp": 4102441322001,
  },
]
`;

exports[`resolveGameState current node is calculated for the current player, not for all players 1`] = `"starter"`;

exports[`resolveGameState resolves an action that gives a global tag as well 1`] = `
Array [
  Object {
    "_partyRef": "party1",
    "_uid": "player1",
    "name": "Player One",
    "tags": Array [
      "BORN ON EARTH",
    ],
  },
  Object {
    "_partyRef": "party1",
    "_uid": "player2",
    "name": "Player Two",
    "tags": Array [],
  },
]
`;

exports[`resolveGameState resolves an action that gives a global tag as well 2`] = `
Array [
  "EARTH DESTROYED",
]
`;

exports[`resolveGameState resolves an action that gives a tag 1`] = `
Array [
  Object {
    "_partyRef": "party1",
    "_uid": "player1",
    "name": "Player One",
    "tags": Array [
      "RAISED BY DIVINE ORDER",
    ],
  },
  Object {
    "_partyRef": "party1",
    "_uid": "player2",
    "name": "Player Two",
    "tags": Array [],
  },
]
`;

exports[`resolveGameState resolves an action that gives a tag 2`] = `"starter-2"`;

exports[`resolveGameState returns the base game state 1`] = `
Object {
  "characters": Array [
    Object {
      "_partyRef": "party1",
      "_uid": "player1",
      "name": "Player One",
      "tags": Array [],
    },
    Object {
      "_partyRef": "party1",
      "_uid": "player2",
      "name": "Player Two",
      "tags": Array [],
    },
  ],
  "currentNode": Object {
    "blocker": false,
    "id": "starter",
    "inGameTimestamp": "2100-01-01 00:00:01",
    "options": Array [
      Object {
        "conditions": Array [],
        "globalConditions": Array [],
        "id": "1",
        "resultingAction": Object {
          "effects": Object {
            "gainTags": Array [
              Object {
                "tag": "BORN ON EARTH",
                "target": "SELF",
              },
              Object {
                "tag": "EARTH DESTROYED",
                "target": "GLOBAL",
              },
            ],
            "loseTags": Array [],
          },
          "id": "1",
          "targetNode": "starter-2",
        },
        "text": "Growing up on Earth was nice… at least before the radiation sickness took Mom and Dad.",
      },
      Object {
        "conditions": Array [],
        "globalConditions": Array [],
        "id": "2",
        "resultingAction": Object {
          "effects": Object {
            "gainTags": Array [
              Object {
                "tag": "RAISED BY DIVINE ORDER",
                "target": "SELF",
              },
            ],
            "loseTags": Array [],
          },
          "id": "2",
          "targetNode": "starter-2",
        },
        "text": "It was fun living here on Fury Station when I was a kid. The Divine Order was strict, but at least they kept the place tidy and safe.",
      },
      Object {
        "conditions": Array [],
        "globalConditions": Array [],
        "id": "3",
        "resultingAction": Object {
          "effects": Object {
            "gainTags": Array [
              Object {
                "tag": "BORN ON A SHIP",
                "target": "SELF",
              },
            ],
            "loseTags": Array [],
          },
          "id": "3",
          "targetNode": "starter-2",
        },
        "text": "I’m glad I was born on a ship. Back then, we were free, the solar wind at our backs.",
      },
    ],
    "text": "Fury Station is a crowded cesspool today. Just like any other day.
The sound of distant PA announcements echo through the hall, and the chatter in dozens of languages merge into a soft, comforting murmur.
The line at the employment office is long enough that you start daydreaming about better times.",
  },
  "globalTags": Array [],
  "log": Array [],
  "player": Object {
    "_partyRef": "party1",
    "_uid": "player1",
    "name": "Player One",
    "tags": Array [],
  },
}
`;
